### ISSUE TITLE
The arm64 image is missing in the docker tags

### ISSUE THREAD

**Issue by strootje** [offen/offen#685]
Hi,

I'd like to be able to run this on a rasberrypi. But there is no arm64 build.
It looks like you are using a cross-compiler for go (xgo). it does have options to add arm64 (`--targets=linux/arm64`)

could it be as simple as this?
.circleci/config.yml
```yml
//...
  build:
    executor: python
    working_directory: ~/offen
    environment:
      DOCKER_LOGIN: offen
      DOCKER_BUILDKIT: '1'
    steps:
      - checkout
      - setup_remote_docker:
          version: 20.10.6
      - run:
          name: Set revision
          command: |
            if [ -z "$CIRCLE_TAG" ]; then
              echo 'export OFFEN_GIT_REVISION=$(git rev-parse --short HEAD)' >> $BASH_ENV
            else
              echo 'export OFFEN_GIT_REVISION=$CIRCLE_TAG' >> $BASH_ENV
            fi
      - run:
          name: Build binaries
          command: |
            if [ -z "$CIRCLE_TAG" ] && [ "${CIRCLE_BRANCH}" != "development" ] && [ "${CIRCLE_BRANCH}" != "master" ]; then
              make build
            else
---           TARGETS=linux/amd64,windows/amd64 make build
+++           TARGETS=linux/amd64,windows/amd64,linux/arm64 make build
              # Darwin builds cannot be statically linked right now
              # so we override the default LDFLAGS value
              TARGETS=darwin/amd64 LDFLAGS='' make build
            fi
      - persist_to_workspace:
          root: ~/offen
          paths:
            - bin/*
//...
```

Would love to hear your thoughts on this

**Comment by m90** [comment]
This would indeed be a great addition. I remember having built a one-off version running on a Pi ages ago, so it should indeed work and should also not be too hard.

What I think would need to happen:
- tell `xgo` to also build a `arm64` binary (like in your diff)
- also tell the Docker build to also build an `linux/arm64` image using the correct binary (this is probably a build arg then instead of a hardcoded path)

In case you want to add this, I am happy to review and merge PRs. Else I'll see if I can get this added soon (it would be a nice thing to have for our upcoming v1).

**Comment by m90** [comment]
I'm sure this can be done in a less hacky way, yet this is a working setup for building a cross-arch Docker image in a sibling repo: https://github.com/offen/docker-volume-backup/blob/30265c14ba967219e7c2dd187b8d71396ea5a5ca/.circleci/config.yml#L37-L52

**Comment by m90** [comment]
It also seems `buildx` would provide us with `TARGETARCH` automatically without having to use a build arg: https://github.com/docker/buildx/issues/574

**Comment by strootje** [comment]
I looked into building an arm64 image for myself prior to this issue. I am currently just using the `golang:1.17-alpine` image. It has support for all the architectures and could theoretically be used to build all archs also. Why do you use `xgo` image over the official `golang` image.

I'm really new to GO, so please forgive the ignorance

**Comment by m90** [comment]
`xgo` is used because Offen uses `go-sqlite` which uses `CGO` which means you need a platform-specific C toolchain (yay for acronyms) if you want statically linked binaries. The `go` tool would only let you build dynamically linked ones when you need to use CGO.

In any case I just tried running `make build TARGETS=linux/arm-7` and got a binary that would run perfectly fine on my Raspbery Pi, so I think this is a solved problem more or less. The task here would just be to have Docker build multiple images per supported architecture and publish them to Docker hub.

**Comment by strootje** [comment]
> `xgo` is used because Offen uses `go-sqlite` which uses `CGO` which means you need a platform-specific C toolchain (yay for acronyms) if you want statically linked binaries. The `go` tool would only let you build dynamically linked ones when you need to use CGO.
Thanks for the explanation. 

I'll have a look at `docker buildx` in the morning

**Comment by m90** [comment]
Keeping this open until we know it's working as intended.

**Comment by m90** [comment]
Looks like your endeavor was a successful one.

After merging #686 on my Raspberry Pi:

```
~ » uname -a                                                                                                                                                                                pi@raspberrypi
Linux raspberrypi 5.10.63-v7l+ #1459 SMP Wed Oct 6 16:41:57 BST 2021 armv7l GNU/Linux
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
~ » docker pull offen/offen:development                                                                                                                                                     pi@raspberrypi
development: Pulling from offen/offen
5480d2ca1740: Pull complete 
9f353405459d: Pull complete 
f500397f935f: Pull complete 
0cb5c27bc29a: Pull complete 
f41aebe758f3: Pull complete 
3a56398afd43: Pull complete 
812a32c5f946: Pull complete 
Digest: sha256:294e0519af304f6707df6e3d298f67762fb65dd97c21b8abb791d29aa220fef8
Status: Downloaded newer image for offen/offen:development
docker.io/offen/offen:development
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
~ » docker run --rm offen/offen:development secret                                                                                                                                          pi@raspberrypi
time="2022-02-15T20:17:04Z" level=info msg="Created 16 bytes secret" secret="ZIdQeT90yuwOFXQKKlXuqA=="
```

And on my laptop:

```
➜  ~ uname -a
Linux fr-xps 5.4.0-99-generic #112-Ubuntu SMP Thu Feb 3 13:50:55 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
➜  ~ docker pull offen/offen:development
development: Pulling from offen/offen
59bf1c3509f3: Already exists 
a4225ece167c: Pull complete 
9fe025ec7dc6: Pull complete 
bf5d62ead980: Pull complete 
9ce3b79174b7: Pull complete 
545bdae29d55: Pull complete 
6ff5a3c6f6a1: Pull complete 
Digest: sha256:294e0519af304f6707df6e3d298f67762fb65dd97c21b8abb791d29aa220fef8
Status: Downloaded newer image for offen/offen:development
docker.io/offen/offen:development
➜  ~ docker run --rm offen/offen:development secret
time="2022-02-15T20:16:22Z" level=info msg="Created 16 bytes secret" secret="JNZUQHvbOJKLzUBhgqZ0xA=="
```

Docker Hub seems content too:

![image](https://user-images.githubusercontent.com/1662740/154142446-276677e9-7e60-423f-a60d-d6aefd49295f.png)

Could you verify that the `development` tag works for your setup as well? In case yes, I can cut a proper release on Thursday so you can pin your version.

**Comment by m90** [comment]
Note to myself: this addition should also be mentioned in the docs.

**Comment by strootje** [comment]
Yes! My pi4 can easily pull a arm64 image now. Thanks for the guidance and the quick responses. They really made this a lot easier :+1:

**Comment by m90** [comment]
This is now released in [v0.7.0](https://github.com/offen/offen/releases/tag/v0.7.0), thanks for working on this!
